docs/training/apis-howto/deepspeed/autotuning.rst

```diff
@@ -0,0 +1,247 @@
+of the ``DeepSpeedEngine`` class need to be wrapped in the :func:`~dsat.dsat_reporting_context` context
+manager. This addition captures the autotuning metrics from each trial and reports the results back
+to the Determined master.
+
+A sketch of example ``dsat`` code with Core API:
+
+.. code:: python
+
+   for op in core_context.searcher.operations():
+      for (inputs, labels) in trainloader:
+          with dsat.dsat_reporting_context(core_context, op): # <-- The new code 
+              outputs = model_engine(inputs)
+              loss = criterion(outputs, labels)
+              model_engine.backward(loss)
+              model_engine.step()
+
+where ``core_context`` is the :class:`~determined.core.Context` instance which was initialized with
+``determined.core.init``. The context manager requires access to both ``core_context`` and the
+current :class:`~determined.core.SearcherOperation` instance (``op``) in order to appropriately report
+results.
+
+
+HuggingFace Trainer
+===================
+
+DeepSpeed Autotune can also be used with the HuggingFace (HF) Trainer and Determined AI's
+:class:`~determined.integrations.huggingface.DetCallback` callback object.
+
+As in the above cases, a ``deepspeed_config`` field  specifying
+the relative path to the DS ``json`` config file must again be added to the
+``hyperparameters`` section of the experiment configuration. Reporting results back to the
+Determined master now requires both using the ``dsat.dsat_reporting_context`` context manager and 
+the ``DetCallback`` callback object listed above.  Additionally, because ``dsat`` performs a search
+over different batch sizes and HuggingFace expects parameters to be specified through command-line
+arguments, an additional helper is needed to create consistent HuggingFace arguments:
+:func:`~dsat.get_hf_args_with_overwrites``.
+
+The key pieces of relevant code from a HuggingFace Trainer script are below.
+
+.. code:: python
+
+  from determined.integrations.huggingface import DetCallback
+  from determined.pytorch import dsat
+  from transformers import HfArgumentParser,Trainer, TrainingArguments,
+
+  hparams = self.context.get_hparams()
+  parser = HfArgumentParser(TrainingArguments)
+  args = sys.argv[1:]
+  args = dsat.get_hf_args_with_overwrites(args, hparams)
+  training_args = parser.parse_args_into_dataclasses(args, look_for_args_file=False)
+
+  det_callback = DetCallback(core_context, ...)
+  trainer = Trainer(args=training_args, ...)
+  with dsat.dsat_reporting_context(core_context, op=det_callback.current_op):
+      train_result = trainer.train(resume_from_checkpoint=checkpoint)
+
+
+Things to note:
+
+* The ``dsat_reporting_context`` context manager shares the same initial
+  :class:`~determined.core.SearcherOperation` as the ``DetCallback`` instance through its
+  ``op=det_callback.current_op`` argument.
+
+* The entire ``train`` method of the HuggingFace trainer is now wrapped in the
+  ``dsat_reporting_context`` context manager.
+
+
```

Comment:
- review of the "HuggingFace" section

**Using DeepSpeed Autotune with HuggingFace Trainer**

You can also use Determined's DeepSpeed Autotune with the HuggingFace (HF) Trainer and Determined's :class:`~determined.integrations.huggingface.DetCallback` callback object to optimize your DeepSpeed parameters.

Similar to the previous cases, you need to add a ``deepspeed_config`` field to the ``hyperparameters`` section of your experiment configuration file, specifying the relative path to the DS ``json`` config file. 

Reporting the results back to the Determined master requires both the ``dsat.dsat_reporting_context`` context manager and the ``DetCallback`` callback object mentioned previously. 

Furthermore, since ``dsat`` performs a search over different batch sizes and HuggingFace expects parameters to be specified as command-line arguments, an additional helper function, :func:`~dsat.get_hf_args_with_overwrites``, is required to create consistent HuggingFace arguments.

Here is an example code snippet from a HuggingFace Trainer script that contains key pieces of relevant code:

.. code:: python
  from determined.integrations.huggingface import DetCallback
  from determined.pytorch import dsat
  from transformers import HfArgumentParser,Trainer, TrainingArguments,
  hparams = self.context.get_hparams()
  parser = HfArgumentParser(TrainingArguments)
  args = sys.argv[1:]
  args = dsat.get_hf_args_with_overwrites(args, hparams)
  training_args = parser.parse_args_into_dataclasses(args, look_for_args_file=False)
  det_callback = DetCallback(core_context, ...)
  trainer = Trainer(args=training_args, ...)
  with dsat.dsat_reporting_context(core_context, op=det_callback.current_op):
      train_result = trainer.train(resume_from_checkpoint=checkpoint)


.. important::

- The ``dsat_reporting_context`` context manager is assigned the same initial :class:`~determined.core.SearcherOperation` as the ``DetCallback`` instance using the ``op=det_callback.current_op`` argument.
- The entire ``train`` method of the HuggingFace trainer is wrapped in the ``dsat_reporting_context`` context manager.